components/server/src/workspace/context-parser-service.ts

```diff
@@ -70,6 +63,87 @@ export class ContextParser {
         return result;
     }
 
+    protected async internalHandleWithoutPrefix(ctx: TraceContext, user: User, nonPrefixedContextURL: string): Promise<WorkspaceContext> {
+        const span = TraceContext.startSpan("ContextParser.internalHandle", ctx);
+        let result: WorkspaceContext | undefined;
+
+        for (const parser of this.allContextParsers) {
+            if (parser.canHandle(user, nonPrefixedContextURL)) {
+                result = await parser.handle({ span }, user, nonPrefixedContextURL);
+                break;
+            }
+        }
+        if (!result) {
+            throw new Error(`Couldn't parse context '${nonPrefixedContextURL}'.`);
+        }
+
+        // TODO: Make the parsers return the context with normalizedContextURL set
+        result.normalizedContextURL = nonPrefixedContextURL;
+        return result;
+    }
+
+    protected buildUpstreamCloneUrl(context: CommitContext): string | undefined {
+        let upstreamCloneUrl: string | undefined = undefined;
+        if (PullRequestContext.is(context) && context.base) {
+            upstreamCloneUrl = context.base.repository.cloneUrl;
+        } else if (context.repository.fork) {
+            upstreamCloneUrl = context.repository.fork.parent.cloneUrl;
+        }
+
+        if (context.repository.cloneUrl === upstreamCloneUrl) {
+            return undefined;
+        }
+        return upstreamCloneUrl;
+    }
+
+    protected async handleMultiRepositoryContext(ctx: TraceContext, user: User, context: WorkspaceContext): Promise<WorkspaceContext> {
+        if (!CommitContext.is(context)) {
+            return context;
+        }
+        const span = TraceContext.startSpan("ContextParser.handleMultiRepositoryContext", ctx);
+        let config = await this.configProvider.fetchConfig({ span }, user, context);
+        let mainRepoContext: WorkspaceContext | undefined;
+        if (config.config.mainConfiguration) {
+            mainRepoContext = await this.internalHandleWithoutPrefix({ span }, user, config.config.mainConfiguration);
+            if (!CommitContext.is(mainRepoContext)) {
+                throw new InvalidGitpodYMLError([`Cannot find main repository '${config.config.mainConfiguration}'.`]);
+            }
+            config = await this.configProvider.fetchConfig({ span }, user, mainRepoContext);
+        }
```

Comment:
> You're probably going to ask if I'm a criminal again (😅)

oops, did I ask that? Sorry, it's an important question to raise. An initializer now can contain multiple git initializers instead of just one and if you don't have access, you cannot initialize a workspace. I have to check, though if we test this properly on prebuilt workspaces.

Reply:
> You're probably going to ask if I'm a criminal again (😅) 
😬 did I ask you that? Sorry, it is an important and very valid question. An initializer now can contain multiple git initializers instead of just one and if you don't have access, you cannot initialize a workspace. I have to check, though if we test this properly on prebuilt workspaces.

Reply:
> An initializer now can contain multiple git initializers instead of just one and if you don't have access, you cannot initialize a workspace.

Makes sense, thanks for explaining.

> I have to check, though if we test this properly on prebuilt workspaces.

👍

> 😬 did I ask you that?

Yes 😂 (but no worries, I wasn't offended 👍). I believe it was during a meeting: you had just proposed a new design idea for something, and I pointed out how it could be potentially abused by nefarious users, and so you asked if I was a criminal because the first thing I say about new ideas is always how they can be abused 😅 (note: I also have strong ethical values that prevent me from abusing things in a bad way, but I do like robust/fail-proof designs, and having a hacker/security-oriented mindset has helped me out many times).